pullman/storageproviders/gcs/provider.go

```diff
@@ -0,0 +1,176 @@
+	if (privateKey == "") != (clientEmail == "") {
+		return nil, errors.New("both private key and client email must be specified for authentication")
+	}
+
+	var creds map[string]string
+	if privateKey != "" && clientEmail != "" {
+		creds = map[string]string{
+			authTypeKey:       authTypeValue,
+			configPrivateKey:  privateKey,
+			configClientEmail: clientEmail,
+		}
+
+		// Only add tokenURI if explicitly specified.
+		tokenUri, _ := pullman.GetString(config, configTokenUri)
+		if tokenUri != "" {
+			creds[configTokenUri] = tokenUri
+		}
+	}
+
+	cl, err := p.gcsDownloaderFactory.newDownloader(log, creds)
+	if err != nil {
+		return nil, err
+	}
+	return &gcsRepositoryClient{
+		gcsclient: cl,
+		log:       log,
+	}, nil
+}
+
+type gcsRepositoryClient struct {
+	gcsclient gcsDownloader
+	log       logr.Logger
+}
+
+// gcsRepository implements RepositoryClient
+var _ pullman.RepositoryClient = (*gcsRepositoryClient)(nil)
+
+func (r *gcsRepositoryClient) Pull(ctx context.Context, pc pullman.PullCommand) error {
+	destDir := pc.Directory
+	targets := pc.Targets
+
+	// Process per-command configuration
+	bucket, ok := pullman.GetString(pc.RepositoryConfig, configBucket)
+	if !ok {
+		return errors.New("required configuration 'bucket' missing from command")
+	}
+
+	// Resolve full paths of objects to download and local paths for the resulting files
+	// Mainly, this means resolving the objects referenced by a "directory" in GCS
+	resolvedTargets := make([]pullman.Target, 0, len(targets))
+	for _, pt := range targets {
+		objPaths, err := r.gcsclient.listObjects(ctx, bucket, pt.RemotePath)
+
+		if err != nil {
+			return fmt.Errorf("unable to list objects in bucket '%s': %w", bucket, err)
+		}
+		r.log.V(1).Info("found objects to download", "path", pt.RemotePath, "count", len(objPaths))
+
+		for _, objPath := range objPaths {
+			localPath := pt.LocalPath
+			relativePath := strings.TrimPrefix(objPath, pt.RemotePath)
+
+			// handle case where the remote path is a single object
+			if relativePath == "" {
+				// allow renaming of the file
+				if localPath != "" && !strings.HasSuffix(localPath, "/") {
+					relativePath = path.Base(localPath)
+					localPath = path.Dir(localPath)
+				} else {
+					relativePath = path.Base(objPath)
+				}
+			}
+
+			filePath, joinErr := util.SecureJoin(destDir, localPath, relativePath)
+			if joinErr != nil {
+				return fmt.Errorf("error joining filepaths '%s' and '%s': %w", pt.LocalPath, relativePath, joinErr)
+			}
+
+			t := pullman.Target{
+				RemotePath: objPath,
+				LocalPath:  filePath,
+			}
+			resolvedTargets = append(resolvedTargets, t)
+		}
+	}
+
+	downloadErr := r.gcsclient.downloadBatch(ctx, bucket, resolvedTargets)
+	if downloadErr != nil {
+		return fmt.Errorf("unable to download objects in bucket '%s': %w", bucket, downloadErr)
+	}
+
+	return nil
+
+}
+
+func init() {
+	p := gcsProvider{
+		gcsDownloaderFactory: gcsClientFactory{},
+	}
+	pullman.RegisterProvider("gs", p)
```

Comment:
Why `gs` rather than `gcs`? (just curious)

Reply:
Ah, here, I was trying to match the scheme of the storage URIs (e.g. `gs://...`). But thinking now that this should be `gcs` to match how the storage provider is referenced in KServe.